ENH: Validation to only allow positive integers for options

Adam-Klaum opened the pull request:
- [x] closes #23348
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I looked at several different options for providing this functionality.  In the end it seemed best to create a a new function that returns a validation function to only allow positive integers.  So now we have is_pos_int available along side the existing is_int and is_text function.  display.max_rows has been modified to use this new validation.  I went through the other integer based display options and tested each with a negative int to see what else might be broken.  display.precision and display.max_columns both had issues when set to a negative number prior to displaying a dataframe.  They have been changed as well.

Adam-Klaum committed:
Created a new validation function (is_pos_int) to allow options to be
set up so that they can be either a None type or a positive int. Updated
the option creation code for display options that failed when set to
negative numbers.

Adam-Klaum committed:
Added positive and negative tests for is_pos_int in test_validation.

Adam-Klaum committed:
Updating whatsnew

Adam-Klaum committed:
Fixing order as per isort

WillAyd commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,34 @@ def inner(x):
+    type_repr = "|".join(map(str, _type))
+
+    def inner(x):
+        if not isinstance(x, _type):
```
I think you could simplify this if you just used a condition like `if not (x is None or isinstance(x, int))`

Adam-Klaum replied:
Agreed,  since the simplified function isn't really dynamic, the latest commit modified is to just be called standalone.

WillAyd commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -1188,6 +1188,7 @@ Other
 - Allow :class:`Index` and :class:`RangeIndex` to be passed to numpy ``min`` and ``max`` functions (:issue:`26125`)
 - Use actual class name in repr of empty objects of a ``Series`` subclass (:issue:`27001`).
 - Bug in :class:`DataFrame` where passing an object array of timezone-aware `datetime` objects would incorrectly raise ``ValueError`` (:issue:`13287`)
+- Added a new option validation function (is_pos_int) to allow only positive integers, and set the display.precision, display.max_rows and display.max_columns options to use it (:issue:`23348`)
```
Since the new function is private no need to mention here. Would want to speak to how its visible to an end user, so something like "Trying to set to a non-positive integer will raise..."

Adam-Klaum replied:
Doc updated with latest commit.

WillAyd commented on pandas/core/config_init.py:
```diff
@@ -319,7 +320,7 @@ def is_terminal():
 
 with cf.config_prefix("display"):
-    cf.register_option("precision", 6, pc_precision_doc, validator=is_int)
+    cf.register_option("precision", 6, pc_precision_doc, validator=is_pos_int())
```
Hmm I don't think this is correct. Isn't this evaluating `is_pos_int()` and returning None, so essentially doing `validator=None`?

Adam-Klaum replied:
This was the correct way to call it, but i didn't really like it as it looked different from the calls to is_int and is_text.  Fixed with the latest commit.

jreback commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,34 @@ def inner(x):
+
+    Returns
+    -------
+    validator - a function of a single argument x , which raises
```
@WillAyd do we check doc-strings here?

Adam-Klaum committed:
Simplified the is_pos_int function to be used standalone rather than
returning an additional function. This also allows it to be used in an
identical fasion to is_int, is_text etc.
Modified the validation calls to is_pos_int to remove ().
Modified the tests using is_pos_int to remove ().
Updated whatsnew documentation to be more user focused.

WillAyd commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,27 @@ def inner(x):
+
+    Parameters
+    ----------
+    value - the value to be checked
```
@jreback comment for returns section is lost in diff so re-commenting here but I'm surprised this isn't failing the CI either

@datapythonista do we ignore private docstrings in the validation? Vaguely recall some comment around that but didn't see code to confirm

Adam-Klaum replied:
Sorry, shall I add the following?

```
Returns
----------
Nothing 
```

datapythonista replied:
@WillAyd yes, we don't validate docstrings of private functions. Ideally we should validate everything, but with all the pending work in the public API that has an impact in the web docs, I don't think we should spend time in fixing the private docstrings yet (and if we don't fix the existing cases, we can't easily validate the new ones).

WillAyd replied:
Thanks for the clarification - do you know where that exclusion happens? Didn't see it in code_checks and couldn't find on quick glance in validate_docstrings.py

datapythonista replied:
Ups sorry, was going to mention and I forgot. It's not that we exclude them, it's that we get the docstrings to consider from the `.rst` files of the API docs. The code here is a good starting point to understand that: https://github.com/pandas-dev/pandas/blob/master/scripts/validate_docstrings.py#L873

TomAugspurger replied:
@Adam-Klaum these need to be updated to follow the doctoring stander. See https://dev.pandas.io/development/contributing.html#updating-a-pandas-docstring

Adam-Klaum committed:
Updated is_pos_int docstring with Returns and Raises

Adam-Klaum commented:
py37_np_dev test is failing  with the error below.  Not sure why.  I notice this is happening on other pull requests as well:

File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.7/site-packages/pytest_nunit/plugin.py", line 276, in <genexpr>
    list(case for case in self.cases.values() if case["outcome"] == "passed")
KeyError: 'outcome'

Adam-Klaum committed:
Clarifying documentation

Adam-Klaum commented:
Is there anything else I need to do on this one? It still shows as Changes requested, but those changes have been integrated into later commits.  I just want to make sure nobody is waiting on me for this one.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -1188,6 +1188,7 @@ Other
 - Allow :class:`Index` and :class:`RangeIndex` to be passed to numpy ``min`` and ``max`` functions (:issue:`26125`)
 - Use actual class name in repr of empty objects of a ``Series`` subclass (:issue:`27001`).
 - Bug in :class:`DataFrame` where passing an object array of timezone-aware `datetime` objects would incorrectly raise ``ValueError`` (:issue:`13287`)
+- Trying to set the display.precision, display.max_rows or display.max_columns using set_option to anything but a None or a positive int will raise a ``ValueError`` (:issue:`23348`)
```
can you move this v0.26.0 docs (pushing shortly)

Adam-Klaum replied:
I don't see a v.0.26.0.rst file out there yet.  Is there a template I should use to create one?

TomAugspurger replied:
Use the 1.0 file.

jreback commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,35 @@ def inner(x):
+    ----------
+    value - the value to be checked
+
+    Returns
```
you don't need a Returns section if its None

jreback commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,35 @@ def inner(x):
+    ----------
+    None
+
+    Raises
```
these have to be the same length as the section

Adam-Klaum committed:
Merge branch 'master' into 23348_max_rows

Adam-Klaum committed:
Removed Returns section from docstring.  Set dash length to be the same
as the Raises keyword in docstring.

Adam-Klaum committed:
Moved whatsnew documentation from v25 to v26

Adam-Klaum committed:
Added documentation to Other section

Adam-Klaum committed:
Merge remote-tracking branch 'origin/master' into 23348_max_rows

jreback commented on doc/source/whatsnew/v0.26.0.rst:
```diff
@@ -0,0 +1,201 @@
+.. _whatsnew_0260:
+
```
this was removed, pls merge master (and remove this file from your push)

Adam-Klaum replied:
Sorry I'm confused.  I was fixing some issues with the v0.26.0.rst file and the index.rst file this morning.  Are you saying you do not want a v0.26.0.rst file now?

jreback replied:
no merge master and put the note in 1.0 exists

Adam-Klaum replied:
Ahh, sorry I misinterpreted "use the 1.0 file" to mean "use the 1.0 file as a template for the v26 file".  I'm on it,  building doc locally now.

Adam-Klaum committed:
Moving comments to v1.0.0 file

TomAugspurger commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -190,6 +190,12 @@ ExtensionArray
+
+Other
+^^^^^
+- Trying to set the display.precision, display.max_rows or display.max_columns using set_option to anything but a None or a positive int will raise a ``ValueError`` (:issue:`23348`)
```
Put the option names like `display.precision` in double-backticks.

set_option -> :func:\`pandas.set_option\` 

a None -> `None`

int -> integer.

TomAugspurger commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,31 @@ def inner(x):
+
+    """
+
+    if not (value is None or isinstance(value, int)):
```
Use `isistance(values, numbers.Integral)`

Can you restructure this a bit?

```python
if value is None:
    return value

elif isinstance(value, numbers.Integral):
    if value < 0:
         raise
    return value

raise ValueError(...)
```

TomAugspurger commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,31 @@ def inner(x):
+    """
+
+    if not (value is None or isinstance(value, int)):
+        msg = "Value must be an instance of <class 'NoneType'>|<class 'int'>"
```
I think most people won't be familiar with NoneType. Should probably say like `'value' must be a positive integer or None`.

Adam-Klaum committed:
Added double ticks around display parameters in whatsnew.
Simplified is_pos_int function and updated tests accordingly
Rewrote is_pos_int docstring to conform to standards

Adam-Klaum committed:
Merge remote-tracking branch 'origin/master' into 23348_max_rows

TomAugspurger commented on pandas/_config/config.py:
```diff
@@ -820,6 +821,31 @@ def inner(x):
+
+    Raises
+    ------
+    ValueError if the value is not equal to a positive integer, 0 or None.
```
I think this should be formatted like

```
Raises
-------
ValueError
    When the value is not a positive integer or None
```

Adam-Klaum replied:
Yes, 0 is allowed, and is necessary for existing internal code.  I think you are right, at this point "is_pos_int" is technically not an accurate name.  I'm fine with "is_nonnegative_int".  The docstring change seems fine to me as well.  I'll make the changes and do another push.

TomAugspurger reviewed (commented):
Small comment on the docstring.

General question: it seems like `0` is allowed? Should we call this `is_nonnegative_int` then, and changes references to "positive int" to "non-negative integer"?

Adam-Klaum committed:
Corrected is_pos_int ValueError message in test suite

Adam-Klaum committed:
Merge remote-tracking branch 'upstream/master' into 23348_max_rows

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -190,6 +190,12 @@ ExtensionArray
+
+Other
+^^^^^
+- Trying to set the ``display.precision``, ``display.max_rows`` or ``display.max_columns`` using set_option to anything but a None or a positive int will raise a ``ValueError`` (:issue:`23348`)
```
use double backticks around None. use ``:meth:`set_option` ``

jreback reviewed (changes requested):
doc-string, ping on green.

Adam-Klaum committed:
Refactored is_pos_int to is_nonnegative_int and updated whatsnew.

Adam-Klaum committed:
Merge remote-tracking branch 'upstream/master' into 23348_max_rows

Adam-Klaum committed:
Fixing import sorting issue

jreback commented:
lgtm. over to you @WillAyd

WillAyd commented:
thanks @Adam-Klaum !
